core/src/nodes/handled_node.rs

```diff
@@ -0,0 +1,261 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use muxing::StreamMuxer;
+use nodes::node::{NodeEvent, NodeStream, Substream};
+use futures::prelude::*;
+use std::io::Error as IoError;
+use Multiaddr;
+
+/// Handler for the substreams of a node.
+///
+/// > Note: When implementing the various methods, don't forget that you have to register the
+/// > task that was the latest to poll and notify it.
+pub trait NodeHandler<TSubstream> {
+    /// Custom event that can be received from the outside.
+    type InEvent;
+    /// Custom event that can be produced by the handler and that will be returned by the swarm.
+    type OutEvent;
+    /// Information about a substream. Can be sent to the handler through a `NodeHandlerEndpoint`,
+    /// and will be passed back in `inject_substream` or `inject_outbound_closed`.
+    type OutboundOpenInfo;
+
+    /// Sends a new substream to the handler.
+    ///
+    /// The handler is responsible for upgrading the substream to whatever protocol it wants.
+    fn inject_substream(&mut self, substream: TSubstream, endpoint: NodeHandlerEndpoint<Self::OutboundOpenInfo>);
+
+    /// Indicates the handler that the inbound part of the muxer has been closed, and that
+    /// therefore no more inbound substream will be produced.
+    fn inject_inbound_closed(&mut self);
+
+    /// Indicates the handler that an outbound substream failed to open because the outbound
+    /// part of the muxer has been closed.
+    fn inject_outbound_closed(&mut self, user_data: Self::OutboundOpenInfo);
+
+    /// Indicates the handler that the multiaddr future has resolved.
+    fn inject_multiaddr(&mut self, multiaddr: Result<Multiaddr, IoError>);
+
+    /// Injects an event coming from the outside in the handler.
+    fn inject_event(&mut self, event: Self::InEvent);
+
+    /// Indicates the node that it should shut down. After that, it is expected that `poll()`
+    /// returns `Ready(None)` as soon as possible.
+    ///
+    /// This method allows an implementation to perform a graceful shutdown of the substreams, and
+    /// send back various events.
+    fn shutdown(&mut self);
+
+    /// Should behave like `Stream::poll()`. Should close if no more event can be produced and the
+    /// node should be closed.
+    fn poll(&mut self) -> Poll<Option<NodeHandlerEvent<Self::OutboundOpenInfo, Self::OutEvent>>, IoError>;
```

Comment:
I don't quite understand the purpose of `poll` in this trait. I can see that client code can either generate an event to ask for a new outbound substream or a custom event, but would it not be easier for clients to be able to call respective methods on some object representing the node? An `open_substream` method would then return `impl Future<Item=TOutboundOpenInfo, Error=...>` and similarly for custom events.

Reply:
I don't want to introduce separate futures, because then you introduce the problem of whether or not polling a future should process the other events that happen on a socket. If not, then things may block forever. If so, then you need to buffer them. Both cases are very magic.